File: NostrVault/app/src/main/java/com/nostrvault/ui/components/CondensedNoteLine.kt
package com.nostrvault.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import coil.request.ImageRequest
import com.nostrvault.data.model.FeedNote
import com.nostrvault.data.model.FeedProfile
import com.nostrvault.data.model.FeedThreadGrouping
import com.nostrvault.ui.theme.*

/**
 * Engagement counts shown beneath a condensed line. Zero fields simply don't
 * draw, so a caller that has no data can pass [CondensedEngagement.NONE].
 * Mirrors iOS `CondensedEngagement`.
 */
data class CondensedEngagement(
    val reactions: Int = 0,
    val reposts: Int = 0,
    val zaps: Int = 0,
    val topEmoji: String? = null,
) {
    val isEmpty: Boolean get() = reactions == 0 && reposts == 0 && zaps == 0

    companion object {
        val NONE = CondensedEngagement()
    }
}

/** How a [CondensedNoteLine] draws its own chrome. Mirrors iOS `CondensedNoteLine.Style`. */
enum class CondensedLineStyle {
    /** Standalone row in the feed: its own bordered card. */
    CARD,

    /** A line inside a thread card, which owns the chrome instead. */
    PLAIN,
}

/** One source for the indent so a line and the row that replaces it when tapped share a left edge. */
fun condensedIndentWidth(depth: Int): Dp =
    (minOf(depth, FeedThreadGrouping.MAX_DEPTH) * 14).dp

/** The rail that ties a reply back to what it answers. */
@Composable
fun ThreadRail(
    isOled: Boolean,
    modifier: Modifier = Modifier,
    themeColor: Color = LocalNostrVaultColors.current.primary,
) {
    Box(
        modifier = modifier
            .width(1.5.dp)
            .fillMaxHeight()
            .padding(end = 8.dp)
            .background(themeColor.copy(alpha = if (isOled) 0.35f else 0.22f)),
    )
}

/**
 * The single condensed representation of a note.
 *
 * Condensed is a property of the feed and nothing else: the feed's condensed
 * and threaded layouts both draw through here, and the thread view is always
 * expanded so a reply is one tap from wherever you landed. Keeping density on
 * one axis is what stops the two surfaces from disagreeing about how dense
 * "condensed" is. Mirrors iOS `CondensedNoteLine.swift`.
 */
@Composable
fun CondensedNoteLine(
    note: FeedNote,
    profile: FeedProfile?,
    profiles: Map<String, FeedProfile> = emptyMap(),
    /** The author to credit when it isn't `note.pubkey` — a bare kind-6 repost shows the original author. */
    displayPubkey: String? = null,
    /** 0 is a root or standalone note; each step indents under a thread rail. */
    depth: Int = 0,
    style: CondensedLineStyle = CondensedLineStyle.CARD,
    /** Draws the purple selection treatment — the focused note in a thread card. */
    isFocused: Boolean = false,
    /** Replies in this thread directly under this note. */
    replyCount: Int = 0,
    /** Text to show instead of `note.content` — an article's title, or the original note behind an empty repost. */
    contentOverride: String? = null,
    mediaURLs: List<String> = emptyList(),
    engagement: CondensedEngagement = CondensedEngagement.NONE,
    showsMediaThumbnail: Boolean = true,
    onProfileClick: (String) -> Unit = {},
    onTap: (() -> Unit)? = null,
    themeColor: Color = LocalNostrVaultColors.current.primary,
    modifier: Modifier = Modifier,
) {
    val isOled = LocalOledMode.current
    val isRoot = depth == 0
    val authorPubkey = displayPubkey ?: note.pubkey
    val displayName = profile?.bestName ?: shortKey(authorPubkey)
    val displayContent = contentOverride ?: note.content

    val avatarSize = if (isRoot) 32.dp else 26.dp
    val nameSize = if (isRoot) 13.sp else 12.sp
    val bodySize = if (isRoot) 14.sp else 13.sp
    val bodyLineLimit = if (isRoot) 3 else 2

    val backgroundColor = when (style) {
        CondensedLineStyle.CARD ->
            if (isFocused) themeColor.copy(alpha = if (isOled) 0.08f else 0.12f) else SecondaryGroupedBg
        CondensedLineStyle.PLAIN ->
            if (isFocused) themeColor.copy(alpha = if (isOled) 0.10f else 0.12f) else Color.Transparent
    }
    val borderColor = when (style) {
        CondensedLineStyle.CARD ->
            if (isFocused) themeColor.copy(alpha = if (isOled) 0.6f else 0.4f)
            else themeColor.copy(alpha = if (isOled) 0.30f else 0.15f)
        CondensedLineStyle.PLAIN ->
            if (isFocused) themeColor.copy(alpha = if (isOled) 0.6f else 0.4f) else Color.Transparent
    }
    val borderWidth = when (style) {
        CondensedLineStyle.CARD -> if (isFocused) 1.5.dp else if (isOled) 1.dp else 0.5.dp
        CondensedLineStyle.PLAIN -> if (isFocused) 1.5.dp else 0.dp
    }
    val cornerRadius = if (style == CondensedLineStyle.CARD) 10.dp else 8.dp

    Row(
        modifier = modifier
            .padding(start = condensedIndentWidth(depth))
            .then(if (onTap != null) Modifier.clickable(onClick = onTap) else Modifier),
        verticalAlignment = Alignment.Top,
    ) {
        if (depth > 0) {
            ThreadRail(isOled = isOled, modifier = Modifier.heightIn(min = 1.dp), themeColor = themeColor)
        }

        Row(
            verticalAlignment = Alignment.Top,
            modifier = Modifier
                .weight(1f)
                .background(backgroundColor, RoundedCornerShape(cornerRadius))
                .border(borderWidth, borderColor, RoundedCornerShape(cornerRadius))
                .padding(
                    horizontal = if (style == CondensedLineStyle.CARD) 12.dp else 8.dp,
                    vertical = if (style == CondensedLineStyle.CARD) 8.dp else 6.dp,
                ),
        ) {
            AvatarImage(
                url = profile?.pictureURL,
                pubkey = authorPubkey,
                size = avatarSize,
                displayName = profile?.bestName,
                modifier = Modifier.clickable { onProfileClick(authorPubkey) },
            )

            Spacer(Modifier.width(8.dp))

            Column(modifier = Modifier.weight(1f)) {
                // Header: name · time · reply count / reply-or-repost marker
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = displayName,
                        color = PrimaryText,
                        fontWeight = FontWeight.SemiBold,
                        fontSize = nameSize,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.weight(1f, fill = false),
                    )
                    Text(
                        text = " · ${formatTimestamp(note.createdAt.time / 1000)}",
                        color = TertiaryText,
                        fontSize = 11.sp,
                        maxLines = 1,
                    )
                    Spacer(Modifier.weight(1f))
                    if (replyCount > 0) {
                        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(3.dp)) {
                            Icon(
                                imageVector = NostrVaultIcons.Chat,
                                contentDescription = null,
                                tint = SecondaryText.copy(alpha = if (isOled) 0.6f else 0.7f),
                                modifier = Modifier.size(9.dp),
                            )
                            Text(
                                text = "$replyCount",
                                color = SecondaryText.copy(alpha = if (isOled) 0.6f else 0.7f),
                                fontSize = 9.sp,
                                fontWeight = FontWeight.SemiBold,
                            )
                        }
                    } else if (note.isReply && depth == 0) {
                        // A reply marker is noise inside a thread — the rail already says it.
                        Icon(
                            imageVector = NostrVaultIcons.Reply,
                            contentDescription = null,
                            tint = themeColor.copy(alpha = 0.7f),
                            modifier = Modifier.size(10.dp),
                        )
                    }
                    if (note.repostedBy != null) {
                        Spacer(Modifier.width(4.dp))
                        Icon(
                            imageVector = NostrVaultIcons.Repost,
                            contentDescription = null,
                            tint = RepostGreen.copy(alpha = 0.7f),
                            modifier = Modifier.size(10.dp),
                        )
                    }
                }

                if (displayContent.isNotBlank()) {
                    Spacer(Modifier.height(2.dp))
                    val plainText = remember(note.id, displayContent, profiles) {
                        NostrMentions.toPlainText(displayContent, profiles, note.mediaURLs.toSet()).replace("\n", " ").trim()
                    }
                    Text(
                        text = plainText,
                        color = SecondaryText,
                        fontSize = bodySize,
                        maxLines = bodyLineLimit,
                        overflow = TextOverflow.Ellipsis,
                        lineHeight = (bodySize.value + 4).sp,
                    )
                }

                if (!engagement.isEmpty) {
                    Spacer(Modifier.height(2.dp))
                    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                        if (engagement.reactions > 0) {
                            EngagementChip(text = engagement.topEmoji ?: "❤️", count = engagement.reactions)
                        }
                        if (engagement.zaps > 0) {
                            EngagementChip(icon = NostrVaultIcons.Zap, tint = ZapOrange, count = engagement.zaps)
                        }
                        if (engagement.reposts > 0) {
                            EngagementChip(icon = NostrVaultIcons.Repost, tint = RepostGreen, count = engagement.reposts)
                        }
                    }
                }
            }

            if (showsMediaThumbnail && mediaURLs.isNotEmpty()) {
                Spacer(Modifier.width(8.dp))
                CondensedMediaThumbnail(mediaURLs, isRoot)
            }
        }
    }
}

@Composable
private fun EngagementChip(
    count: Int,
    text: String? = null,
    icon: androidx.compose.ui.graphics.vector.ImageVector? = null,
    tint: Color = SecondaryText,
) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(2.dp)) {
        when {
            icon != null -> Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(8.dp))
            text != null -> Text(text, fontSize = 9.sp)
        }
        Text("$count", color = SecondaryText, fontSize = 9.sp, fontWeight = FontWeight.SemiBold)
    }
}

@Composable
private fun CondensedMediaThumbnail(mediaURLs: List<String>, isRoot: Boolean) {
    val context = LocalContext.current
    val size = if (isRoot) 80.dp else 56.dp
    Box {
        AsyncImage(
            model = ImageRequest.Builder(context)
                .data(mediaURLs.first())
                .size(if (isRoot) 160 else 112)
                .crossfade(false)
                .build(),
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier
                .size(size)
                .clip(RoundedCornerShape(6.dp))
                .background(TertiaryGroupedBg),
        )
        if (mediaURLs.size > 1) {
            Text(
                text = "+${mediaURLs.size - 1}",
                color = Color.White,
                fontSize = 10.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier
                    .align(Alignment.BottomEnd)
                    .padding(4.dp)
                    .background(Color.Black.copy(alpha = 0.7f), RoundedCornerShape(50))
                    .padding(horizontal = 6.dp, vertical = 3.dp),
            )
        }
    }
}

private fun shortKey(key: String): String =
    if (key.length < 12) key else "npub…" + key.takeLast(6)

File: HavenApp/HavenApp/Views/Components/CondensedNoteLine.swift
import SwiftUI

/// Engagement counts shown beneath a condensed line. Zero fields simply don't
/// draw, so a caller that has no data can pass `.none`.
struct CondensedEngagement: Equatable {
    var reactions: Int = 0
    var reposts: Int = 0
    var zaps: Int = 0
    var topEmoji: String? = nil

    static let none = CondensedEngagement()

    var isEmpty: Bool { reactions == 0 && reposts == 0 && zaps == 0 }
}

/// The single condensed representation of a note.
///
/// Condensed is a property of the feed and nothing else: the feed's condensed
/// and threaded layouts both draw through here, and the thread view is always
/// expanded so a reply is one tap from wherever you landed. Keeping density on
/// one axis is what stops the two surfaces from disagreeing about how dense
/// "condensed" is.
struct CondensedNoteLine: View {
    enum Style {
        /// Standalone row in the feed: its own bordered card.
        case card
        /// A line inside a thread card, which owns the chrome instead.
        case plain
    }

    let note: FeedNote
    let profile: FeedProfile?
    /// The author to credit when it isn't `note.pubkey` — a bare kind-6 repost
    /// shows the original author, not the reposter.
    var displayPubkey: String? = nil

    /// 0 is a root or standalone note; each step indents under a thread rail.
    var depth: Int = 0
    var style: Style = .card
    /// Draws the purple selection treatment — the focused note in a thread card.
    var isFocused: Bool = false
    /// Replies in this thread directly under this note.
    var replyCount: Int = 0
    /// Text to show instead of `note.content` — an article's title, or the
    /// original note's body behind an empty repost.
    var contentOverride: String? = nil
    /// Media for the thumbnail. Callers resolve reposts before passing it.
    var mediaURLs: [URL] = []
    var engagement: CondensedEngagement = .none
    var showsMediaThumbnail: Bool = true

    var onProfile: ((String) -> Void)? = nil
    var onTap: (() -> Void)? = nil

    @EnvironmentObject private var configService: ConfigService

    private var isOLED: Bool { configService.config.useOLED }
    private var isRoot: Bool { depth == 0 }

    /// The root anchors the line; replies step in under it. Capped by the
    /// grouper so a deep argument can never indent content off-screen.
    private var indentWidth: CGFloat { CondensedNoteLine.indentWidth(forDepth: depth) }

    /// One source for the indent so a line and the full row that replaces it
    /// when tapped sit on the same left edge.
    static func indentWidth(forDepth depth: Int) -> CGFloat {
        CGFloat(min(depth, FeedThreadGrouping.maxDepth)) * 14
    }

    /// The rail that ties a reply back to what it answers, drawn standalone so
    /// an expanded row can keep the same thread line a condensed one has.
    static func rail(isOLED: Bool) -> some View {
        Rectangle()
            .fill(Color.havenPurple.opacity(isOLED ? 0.35 : 0.22))
            .frame(width: 1.5)
            .padding(.trailing, 8)
            .accessibilityHidden(true)
    }

    private var avatarSize: CGFloat { CondensedNoteLine.avatarSize(forDepth: depth) }

    /// One source for the avatar size so a line and the full row that
    /// replaces it when tapped can match it exactly — opening a line adds an
    /// action bar, not a size change.
    static func avatarSize(forDepth depth: Int) -> CGFloat {
        depth == 0 ? 32 : 26
    }
    private var nameSize: CGFloat { isRoot ? 13 : 12 }
    private var bodySize: CGFloat { isRoot ? 14 : 13 }
    private var bodyLineLimit: Int { isRoot ? 3 : 2 }

    private var displayContent: String {
        contentOverride ?? note.content
    }

    var body: some View {
        HStack(alignment: .top, spacing: 0) {
            if depth > 0 {
                threadRail
            }

            HStack(alignment: .top, spacing: 8) {
                AvatarView(url: profile?.pictureURL, pubkey: authorPubkey, size: avatarSize)
                    .contentShape(Circle())
                    .onTapGesture { onProfile?(authorPubkey) }
                    .accessibilityLabel(Text("Profile of \(displayName)"))

                VStack(alignment: .leading, spacing: 2) {
                    headerRow
                    bodyText
                    engagementRow
                }

                Spacer(minLength: 8)

                if showsMediaThumbnail, let firstMedia = mediaURLs.first {
                    mediaThumbnail(firstMedia)
                }
            }
            .padding(.horizontal, style == .card ? 12 : 8)
            .padding(.vertical, style == .card ? 8 : 6)
            .background(rowBackground)
            .overlay(rowBorder)
        }
        .padding(.leading, indentWidth)
        .contentShape(Rectangle())
        .onTapGesture { onTap?() }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(Text(accessibilityLabel))
        .accessibilityAddTraits(onTap == nil ? [] : .isButton)
    }

    // MARK: - Pieces

    /// The vertical line that ties a reply back to what it answers.
    private var threadRail: some View {
        CondensedNoteLine.rail(isOLED: isOLED)
    }

    private var headerRow: some View {
        HStack(spacing: 4) {
            Text(displayName)
                .font(.appSystem(size: nameSize, weight: .semibold))
                .foregroundColor(.white.opacity(isRoot ? 1.0 : (isOLED ? 0.92 : 0.95)))
                .lineLimit(1)

            if let nip05 = profile?.nip05, !nip05.isEmpty {
                Image(systemName: "checkmark.seal.fill")
                    .font(.appSystem(size: 9))
                    .foregroundColor(Color(red: 0.2, green: 0.8, blue: 0.6))
            }

            Text("· \(CondensedNoteLine.relativeTime(note.createdAt))")
                .font(.appSystem(size: 11))
                .foregroundColor(.secondary)

            Spacer(minLength: 4)

            if replyCount > 0 {
                HStack(spacing: 3) {
                    Image(systemName: "text.bubble")
                        .font(.appSystem(size: 9, weight: .medium))
                    Text("\(replyCount)")
                        .font(.appSystem(size: 9, weight: .semibold, design: .monospaced))
                }
                .foregroundColor(.secondary.opacity(isOLED ? 0.6 : 0.7))
            }

            // A reply marker is noise inside a thread — the rail already says it.
            if note.isReply && depth == 0 && replyCount == 0 {
                Image(systemName: "arrowshape.turn.up.left.fill")
                    .font(.appSystem(size: 10))
                    .foregroundColor(Color.havenPurple.opacity(0.7))
            }
            if note.repostedBy != nil {
                Image(systemName: "arrow.2.squarepath")
                    .font(.appSystem(size: 10))
                    .foregroundColor(.green.opacity(0.7))
            }
        }
    }

    @ViewBuilder
    private var bodyText: some View {
        if !displayContent.isEmpty {
            Text(NostrContentFormatter.resolveMentionsPlainText(displayContent))
                .font(.appSystem(size: bodySize))
                .foregroundColor(.white.opacity(isRoot ? 1.0 : (isOLED ? 0.8 : 0.85)))
                .lineLimit(bodyLineLimit)
                .lineSpacing(1)
                .multilineTextAlignment(.leading)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    @ViewBuilder
    private var engagementRow: some View {
        if !engagement.isEmpty {
            HStack(spacing: 6) {
                if engagement.reactions > 0 {
                    countChip(
                        symbol: .emoji(engagement.topEmoji ?? "❤️"),
                        count: engagement.reactions,
                        tint: .secondary
                    )
                }
                if engagement.zaps > 0 {
                    countChip(symbol: .system("bolt.fill"), count: engagement.zaps, tint: .orange)
                }
                if engagement.reposts > 0 {
                    countChip(symbol: .system("arrow.2.squarepath"), count: engagement.reposts, tint: .green)
                }
            }
            .padding(.top, 2)
        }
    }

    private enum ChipSymbol {
        case emoji(String)
        case system(String)
    }

    private func countChip(symbol: ChipSymbol, count: Int, tint: Color) -> some View {
        HStack(spacing: 2) {
            switch symbol {
            case .emoji(let value):
                Text(value).font(.appSystem(size: 9))
            case .system(let name):
                Image(systemName: name)
                    .font(.appSystem(size: 8, weight: .bold))
                    .foregroundColor(tint)
            }
            Text("\(count)")
                .font(.appSystem(size: 9, weight: .semibold, design: .monospaced))
                .foregroundColor(.secondary)
        }
    }

    private func mediaThumbnail(_ url: URL) -> some View {
        ZStack(alignment: .bottomTrailing) {
            FeedMediaView(url: url, isThumbnail: true)
                .frame(width: isRoot ? 80 : 56, height: isRoot ? 80 : 56)
                .aspectRatio(1, contentMode: .fill)
                .clipShape(RoundedRectangle(cornerRadius: 6))

            if mediaURLs.count > 1 {
                Text("+\(mediaURLs.count - 1)")
                    .font(.appSystem(size: 10, weight: .bold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 6)
                    .padding(.vertical, 3)
                    .background(Color.black.opacity(0.7))
                    .clipShape(Capsule())
                    .padding(4)
            }
        }
        .accessibilityHidden(true)
    }

    // MARK: - Chrome

    @ViewBuilder
    private var rowBackground: some View {
        switch style {
        case .card:
            RoundedRectangle(cornerRadius: 10)
                .fill(isFocused ? Color.havenPurple.opacity(isOLED ? 0.08 : 0.12) : Color.platformSecondaryGroupedBackground)
                .overlay(
                    RoundedRectangle(cornerRadius: 10)
                        .fill(Color.havenPurple.opacity(0.015))
                )
        case .plain:
            // Inside a thread card the rows share one surface; only the focused
            // line is tinted, so the conversation reads as a block instead of a
            // stack of boxes.
            RoundedRectangle(cornerRadius: 8)
                .fill(isFocused ? Color.havenPurple.opacity(isOLED ? 0.10 : 0.12) : Color.clear)
        }
    }

    @ViewBuilder
    private var rowBorder: some View {
        switch style {
        case .card:
            RoundedRectangle(cornerRadius: 10)
                .stroke(
                    isFocused ? Color.havenPurple.opacity(isOLED ? 0.6 : 0.4) : Color.havenPurple.opacity(isOLED ? 0.30 : 0.15),
                    lineWidth: isFocused ? 1.5 : (isOLED ? 1.0 : 0.5)
                )
        case .plain:
            if isFocused {
                RoundedRectangle(cornerRadius: 8)
                    .stroke(Color.havenPurple.opacity(isOLED ? 0.6 : 0.4), lineWidth: 1.5)
            }
        }
    }

    // MARK: - Text

    private var authorPubkey: String { displayPubkey ?? note.pubkey }

    private var displayName: String {
        profile?.bestName ?? CondensedNoteLine.shortKey(authorPubkey)
    }

    private var accessibilityLabel: String {
        var parts = [displayName, CondensedNoteLine.relativeTime(note.createdAt)]
        if depth > 0 { parts.append("reply, level \(depth)") }
        if !displayContent.isEmpty {
            parts.append(NostrContentFormatter.resolveMentionsPlainText(displayContent))
        }
        if replyCount > 0 {
            parts.append("\(replyCount) \(replyCount == 1 ? "reply" : "replies")")
        }
        if !mediaURLs.isEmpty {
            parts.append(mediaURLs.count == 1 ? "1 attachment" : "\(mediaURLs.count) attachments")
        }
        return parts.joined(separator: ", ")
    }

    static func shortKey(_ key: String) -> String {
        guard key.count >= 12 else { return key }
        return "npub…" + String(key.suffix(6))
    }

    static func relativeTime(_ date: Date) -> String {
        let diff = Date().timeIntervalSince(date)
        switch diff {
        case ..<60:         return "now"
        case ..<3600:       return "\(Int(diff / 60))m"
        case ..<86400:      return "\(Int(diff / 3600))h"
        case ..<604800:     return "\(Int(diff / 86400))d"
        default:
            let fmt = DateFormatter()
            fmt.dateFormat = Calendar.current.isDate(date, equalTo: Date(), toGranularity: .year)
                ? "MMM d" : "MMM d, yyyy"
            return fmt.string(from: date)
        }
    }
}

/// The card a condensed conversation sits in — one surface for the whole
/// thread, so replies read as a block instead of a stack of boxes.
struct ThreadCardBackground: ViewModifier {
    var isOLED: Bool = ConfigService.shared.config.useOLED

    func body(content: Content) -> some View {
        content
            .padding(.vertical, 8)
            .padding(.horizontal, 6)
            .background(
                ZStack {
                    Color.platformSecondaryGroupedBackground
                    Color.havenPurple.opacity(0.015)
                }
            )
            .cornerRadius(12)
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(
                        Color.havenPurple.opacity(isOLED ? 0.30 : 0.15),
                        lineWidth: isOLED ? 1.0 : 0.5
                    )
            )
    }
}

extension View {
    /// Wrap a run of `CondensedNoteLine`s as a single conversation card.
    func threadCard() -> some View {
        modifier(ThreadCardBackground())
    }
}
